File: 905-sort-array-by-parity/905-sort-array-by-parity.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArrayByParity = function(nums) {
  nums.sort((a, b) => {
    const aIsOdd = a % 2
    const bIsOdd = b % 2
    if (!aIsOdd && bIsOdd) return -1
    else if (aIsOdd && !bIsOdd) return 1
    else return a - b
  }) 
  return nums
};

File: 905-sort-array-by-parity/905-sort-array-by-parity.go
func sortArrayByParity(nums []int) []int {
  n := len(nums)
  fast, slow := 0, 0
  for slow < n && fast < n {
    for ; slow < n; slow ++ {
      if nums[slow] % 2 == 1 {
        break
      }
    }
    fast = slow + 1
    for ; fast < n; fast ++ {
      if nums[fast] % 2 == 0 {
        break
      }
    }
    if slow < n && fast < n {
      nums[slow], nums[fast] = nums[fast], nums[slow]
    }
    slow ++
  }
  return nums
}